File: chisel_js/src/chisel/checker/services/agent_skills.ts
import { SkillTarget, type SkillInstallResult } from "chisel/checker/models/agent_skill";
import {
  cpSync,
  existsSync,
  mkdirSync,
  mkdtempSync,
  readdirSync,
  readFileSync,
  renameSync,
  rmSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { basename, dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const TARGET_DIRS: Record<SkillTarget, string> = {
  [SkillTarget.CODEX]: ".agents/skills",
  [SkillTarget.CLAUDE]: ".claude/skills",
  [SkillTarget.OPENCODE]: ".opencode/skills",
};

export interface ISkillInstaller {
  availableSkills(): string[];
  install(
    projectPath: string,
    target: SkillTarget,
    options?: {
      readonly skillNames?: string[];
      readonly overwrite?: boolean;
      readonly dryRun?: boolean;
    },
  ): SkillInstallResult[];
  targetDir(target: SkillTarget): string;
}

export class SkillInstaller {
  constructor(private readonly sourceDir?: string) {}

  availableSkills(): string[] {
    return this.skillDirs().map(path => path.name);
  }

  install(
    projectPath: string,
    target: SkillTarget,
    options: {
      readonly skillNames?: string[];
      readonly overwrite?: boolean;
      readonly dryRun?: boolean;
    } = {},
  ): SkillInstallResult[] {
    const root = resolve(projectPath);
    const selected = new Set(options.skillNames ?? []);
    const available = new Map(this.skillDirs().map(path => [path.name, path.path]));
    const results: SkillInstallResult[] = [];

    for (const missing of [...selected].filter(name => !available.has(name)).sort()) {
      const destination = join(root, TARGET_DIRS[target], missing);
      results.push({
        name: missing,
        source: "",
        destination,
        status: "skipped_missing",
        reason: "No bundled skill with this name exists.",
      });
    }

    for (const [name, source] of [...available.entries()].sort()) {
      if (selected.size && !selected.has(name)) continue;
      this.validateSkill(source, name);
      const destination = join(root, TARGET_DIRS[target], name);
      if (existsSync(destination) && !options.overwrite) {
        results.push({
          name,
          source,
          destination,
          status: "skipped_exists",
          reason: "Destination already exists. Pass --overwrite to replace it.",
        });
        continue;
      }
      if (options.dryRun) {
        results.push({
          name,
          source,
          destination,
          status: "would_install",
          reason: "Dry run; no files were written.",
        });
        continue;
      }
      this.copySkill(source, destination, options.overwrite ?? false);
      results.push({
        name,
        source,
        destination,
        status: "installed",
        reason: "Installed skill.",
      });
    }

    return results;
  }

  targetDir(target: SkillTarget): string {
    return TARGET_DIRS[target];
  }

  private copySkill(source: string, destination: string, overwrite: boolean): void {
    const parent = dirname(destination);
    mkdirSync(parent, { recursive: true });
    const skillName = basename(destination) || "skill";
    const tempParent = mkdtempSync(join(parent || tmpdir(), `.${skillName}.`));
    const tempDestination = join(tempParent, skillName);
    cpSync(source, tempDestination, {
      recursive: true,
      filter: path => !path.endsWith("__pycache__") && !path.endsWith(".pyc"),
    });
    if (existsSync(destination)) {
      if (overwrite) {
        rmSync(destination, { recursive: true, force: true });
      } else {
        rmSync(tempParent, { recursive: true, force: true });
        return;
      }
    }
    renameSync(tempDestination, destination);
    rmSync(tempParent, { recursive: true, force: true });
  }

  private skillDirs(): Array<{ readonly name: string; readonly path: string }> {
    const source = this.resolveSourceDir();
    return readdirSync(source, { withFileTypes: true })
      .filter(entry => entry.isDirectory() && existsSync(join(source, entry.name, "SKILL.md")))
      .map(entry => ({ name: entry.name, path: join(source, entry.name) }));
  }

  private resolveSourceDir(): string {
    if (this.sourceDir) return this.sourceDir;
    const moduleDir = dirname(fileURLToPath(import.meta.url));
    const candidates = [
      join(moduleDir, "../bundled_skills"),
      join(moduleDir, "../../bundled_skills"),
      join(moduleDir, "../../skills"),
      join(moduleDir, "../../../../../skills"),
    ];
    for (const candidate of candidates) {
      if (existsSync(candidate)) return candidate;
    }
    throw new Error("Bundled Chisel skills could not be found.");
  }

  private validateSkill(skillDir: string, expectedName: string): void {
    const frontmatter = this.frontmatter(join(skillDir, "SKILL.md"));
    if (frontmatter.name !== expectedName) {
      throw new Error(`${join(skillDir, "SKILL.md")} must declare name: ${expectedName}`);
    }
    if (!frontmatter.description) {
      throw new Error(`${join(skillDir, "SKILL.md")} must declare a description.`);
    }
  }

  private frontmatter(skillFile: string): Record<string, string> {
    const lines = readFileSync(skillFile, "utf-8").split(/\r?\n/);
    if (!lines.length || lines[0].trim() !== "---") {
      throw new Error(`${skillFile} must start with YAML frontmatter.`);
    }
    const data: Record<string, string> = {};
    for (const line of lines.slice(1)) {
      if (line.trim() === "---") return data;
      if (!line.includes(":")) continue;
      const [key, ...valueParts] = line.split(":");
      data[key.trim()] = valueParts.join(":").trim().replace(/^"|"$/g, "");
    }
    throw new Error(`${skillFile} frontmatter is not closed.`);
  }
}

File: chisel_py/src/chisel/checker/services/agent_skills.py
from __future__ import annotations

import shutil
import tempfile
from dataclasses import dataclass
from importlib import resources
from pathlib import Path
from typing import Protocol, runtime_checkable

from chisel.checker.models.agent_skill import SkillInstallResult, SkillTarget


_TARGET_DIRS: dict[SkillTarget, Path] = {
    SkillTarget.CODEX: Path(".agents") / "skills",
    SkillTarget.CLAUDE: Path(".claude") / "skills",
    SkillTarget.OPENCODE: Path(".opencode") / "skills",
}


@runtime_checkable
class ISkillInstaller(Protocol):
    def available_skills(self) -> list[str]: ...

    def install(
        self,
        project_path: Path,
        target: SkillTarget,
        skill_names: list[str] | None = None,
        overwrite: bool = False,
        dry_run: bool = False,
    ) -> list[SkillInstallResult]: ...

    def target_dir(self, target: SkillTarget) -> Path: ...


@dataclass(slots=True)
class SkillInstaller:
    _source_dir: Path | None = None

    def available_skills(self) -> list[str]:
        return [path.name for path in self._skill_dirs()]

    def install(
        self,
        project_path: Path,
        target: SkillTarget,
        skill_names: list[str] | None = None,
        overwrite: bool = False,
        dry_run: bool = False,
    ) -> list[SkillInstallResult]:
        project_path = project_path.resolve()
        selected = set(skill_names or [])
        available = {path.name: path for path in self._skill_dirs()}

        results: list[SkillInstallResult] = []
        for missing in sorted(selected - available.keys()):
            destination = project_path / _TARGET_DIRS[target] / missing
            results.append(
                SkillInstallResult(
                    name=missing,
                    source="",
                    destination=str(destination),
                    status="skipped_missing",
                    reason="No bundled skill with this name exists.",
                )
            )

        for name, source in sorted(available.items()):
            if selected and name not in selected:
                continue
            self._validate_skill(source, name)
            destination = project_path / _TARGET_DIRS[target] / name
            if destination.exists() and not overwrite:
                results.append(
                    SkillInstallResult(
                        name=name,
                        source=str(source),
                        destination=str(destination),
                        status="skipped_exists",
                        reason="Destination already exists. Pass --overwrite to replace it.",
                    )
                )
                continue
            if dry_run:
                results.append(
                    SkillInstallResult(
                        name=name,
                        source=str(source),
                        destination=str(destination),
                        status="would_install",
                        reason="Dry run; no files were written.",
                    )
                )
                continue
            self._copy_skill(source, destination, overwrite=overwrite)
            results.append(
                SkillInstallResult(
                    name=name,
                    source=str(source),
                    destination=str(destination),
                    status="installed",
                    reason="Installed skill.",
                )
            )

        return results

    def target_dir(self, target: SkillTarget) -> Path:
        return _TARGET_DIRS[target]

    def _copy_skill(
        self, source: Path, destination: Path, overwrite: bool
    ) -> None:
        destination.parent.mkdir(parents=True, exist_ok=True)
        temp_parent = destination.parent
        with tempfile.TemporaryDirectory(
            prefix=f".{destination.name}.", dir=temp_parent
        ) as tmp:
            temp_destination = Path(tmp) / destination.name
            shutil.copytree(
                source,
                temp_destination,
                ignore=shutil.ignore_patterns("__pycache__", "*.pyc"),
            )
            if destination.exists():
                if overwrite:
                    shutil.rmtree(destination)
                else:
                    return
            temp_destination.rename(destination)

    def _skill_dirs(self) -> list[Path]:
        source_dir = self._resolve_source_dir()
        return [
            path
            for path in source_dir.iterdir()
            if path.is_dir() and (path / "SKILL.md").is_file()
        ]

    def _resolve_source_dir(self) -> Path:
        if self._source_dir is not None:
            return self._source_dir

        package_source = resources.files("chisel").joinpath("bundled_skills")
        if package_source.is_dir():
            return Path(str(package_source))

        repo_source = Path(__file__).resolve().parents[5] / "skills"
        if repo_source.is_dir():
            return repo_source

        raise FileNotFoundError("Bundled Chisel skills could not be found.")

    def _validate_skill(self, skill_dir: Path, expected_name: str) -> None:
        frontmatter = self._frontmatter(skill_dir / "SKILL.md")
        name = frontmatter.get("name")
        description = frontmatter.get("description")
        if name != expected_name:
            raise ValueError(
                f"{skill_dir / 'SKILL.md'} must declare name: {expected_name}"
            )
        if not description:
            raise ValueError(
                f"{skill_dir / 'SKILL.md'} must declare a description."
            )

    def _frontmatter(self, skill_file: Path) -> dict[str, str]:
        lines = skill_file.read_text(encoding="utf-8").splitlines()
        if not lines or lines[0].strip() != "---":
            raise ValueError(f"{skill_file} must start with YAML frontmatter.")
        data: dict[str, str] = {}
        for line in lines[1:]:
            if line.strip() == "---":
                return data
            if ":" not in line:
                continue
            key, value = line.split(":", 1)
            data[key.strip()] = value.strip().strip('"')
        raise ValueError(f"{skill_file} frontmatter is not closed.")
